chain/epoch-manager/src/validator_selection.rs

```diff
@@ -33,49 +32,83 @@ pub fn proposals_to_epoch_info(
+    // Returns unselected proposals, validator lists for all roles and stake
+    // threshold to become a validator.
+    let (unselected_proposals, chunk_producers, block_producers, chunk_validators, threshold) =
+        if checked_feature!("stable", StatelessValidationV0, next_version) {
+            let min_stake_ratio = {
+                let rational = epoch_config.validator_selection_config.minimum_stake_ratio;
+                Ratio::new(*rational.numer() as u128, *rational.denom() as u128)
+            };
+
+            let mut chunk_producer_proposals = order_proposals(proposals.values().cloned());
             let (chunk_producers, cp_stake_threshold) = select_chunk_producers(
                 &mut chunk_producer_proposals,
-                max_cp_selected,
+                epoch_config.validator_selection_config.num_chunk_producer_seats as usize,
                 min_stake_ratio,
                 shard_ids.len() as NumShards,
-                current_version,
+                next_version,
+            );
+
+            let mut block_producer_proposals = order_proposals(proposals.values().cloned());
+            let (block_producers, bp_stake_threshold) = select_block_producers(
+                &mut block_producer_proposals,
+                epoch_config.num_block_producer_seats as usize,
+                min_stake_ratio,
+                next_version,
+            );
+
+            let mut chunk_validator_proposals = order_proposals(proposals.into_values());
+            let (chunk_validators, cv_stake_threshold) = select_validators(
+                &mut chunk_validator_proposals,
+                epoch_config.validator_selection_config.num_chunk_validator_seats as usize,
+                min_stake_ratio,
+                next_version,
+            );
+
+            // Note that if there are too few validators and too many shards,
+            // assigning chunk producers to shards is more aggressive, so it
+            // is not enough to iterate over chunk validators.
+            // So unfortunately we have to look over all roles to get unselected
+            // proposals.
+            // TODO: getting unselected proposals must be simpler. For example,
+            // we can track all roles assign to each validator in some structure
+            // and then return all validators which don't have any role.
+            let max_validators_for_role = cmp::max(
+                chunk_producers.len(),
+                cmp::max(block_producers.len(), chunk_validators.len()),
             );
-            (chunk_producer_proposals, chunk_producers, cp_stake_threshold)
+            let unselected_proposals = if chunk_producers.len() == max_validators_for_role {
+                chunk_producer_proposals
+            } else if block_producers.len() == max_validators_for_role {
+                block_producer_proposals
+            } else {
+                chunk_validator_proposals
+            };
```

Comment:
Can you explain what are the unselected proposals and why is it the ones with max validator role? In a typical configuration we're going to have the most chunk validators. Why do we make all chunk validator proposals "unselected"?

Reply:
This follows current implementation style. When one calls `select_validators`, passed proposals ( e.g. `chunk_validator_proposals`) are mutated: if proposal is selected, it is popped out of set, so the only ones remaining are unselected.

Then, I use the fact that all roles are selected by stake, so the unselected proposals will come from the role with maximal amount of validators. All other proposals will have at least one role assigned.

This is not ideal but I don't want to tweak this part of implementation as well for now.